### PR TITLE
Improve cmake

### PR DESCRIPTION
**Pull request by martinhruska**
This PR modifies header files and CMake files to
- enable includion of mata without need to include simlib
- to create one static library wrapping libmata and all its dependencies
- add install to CMake and Make so after `make install` it it possible to use Mata by adding the following to user's project:
```
find_library(LIBMATA mata)
target_link_libraries(demo ${LIBMATA})
```

### CODE REVIEW

File: merge_static_libraries.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+do
+  echo "SRC $i"
+  cp ${i} ${TMP_DIR}
+done
+
+echo "LIBS copied"
+(
+	cd ${TMP_DIR}
+	for i in *.a
+  do
```
Comment (tfiedor) [inline merge_static_libraries.sh]: Just minor thing, but indent here sounds weird.

**Comment by martinhruska** [comment]
I have very limited knowledge of CMake so this PR is rather a prototype. I appreciate your feedback about possible improvements.

**Review by tfiedor — APPROVED** [review]
Seems OK, though I wonder, if we could test the installation in github action somehow?

File: merge_static_libraries.sh (hunk 1 above)
Comment (martinhruska) [reply merge_static_libraries.sh]: solved.

**Comment by martinhruska** [comment]
Well I used the following dummy examples:
```
#include <mata/nfa.hh>
#include <iostream>

int main() {
  Mata::Nfa::Nfa aut;
  aut.add_state();
  std::cout << "Hello, Mata!" << std::endl;
  return 0;
}
```
with the following CMake:
```
cmake_minimum_required(VERSION 3.3)
project(Demo)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
add_executable(demo demo.cc)
find_library(LIBMATA mata)
target_link_libraries(demo ${LIBMATA})
```

Maybe, we can create somethin more complex, add it to a special folder and run cmake inside it. However, we need to run `make install` over mata. I am not sure if github actions allows us to export files to the system folders.

**Comment by vhavlena** [comment]
I have a problem with RE2 (missing symbols from RE2 during linking). A working solution for me is to copy `libre2.a` to the system folder during the installation of mata and then in a testing project simply use `find_library(LIBRE2 re2)` together with linking `${LIBRE2}`.

**Review by vhavlena — APPROVED** [review]

**Comment by vhavlena** [comment]
Perfectly works even in z3.

**Comment by martinhruska** [comment]
Great, then if @tfiedor kindly fix the bindings I'will merge it asap.

**Review by Adda0 — APPROVED** [review]
LGTM. This should hopefully help a lot.

**Comment by tfiedor** [comment]
> Great, then if @tfiedor kindly fix the bindings I'will merge it asap.

OMG, ok, will do it, though this change shouldn't have broken it.

**Comment by martinhruska** [comment]
Actually, it was needed to move methods related to simulation from nfa.hh to algorithms since we don't want to export include of external library with basic NFA interface.

**Comment by Adda0** [comment]
> > Great, then if @tfiedor kindly fix the bindings I'will merge it asap.
> 
> OMG, ok, will do it, though this change shouldn't have broken it.

It should have. `compute_relation()` function was moved from `nfa.hh` to `nfa-algorithms.hh`. Therefore, we try to bind to undefined method `compute_relation()` in `nfa.hh` in Python binding.

**Comment by tfiedor** [comment]
> Actually, it was needed to move methods related to simulation from nfa.hh to algorithms since we don't want to export include of external library with basic NFA interface.

Yeah, it's OK. Can you please write me where was `compute_relation` moved? From `Nfa::` to `Nfa::Algorithms`?

**Comment by tfiedor** [comment]
> > > Great, then if @tfiedor kindly fix the bindings I'will merge it asap.
> > 
> > 
> > OMG, ok, will do it, though this change shouldn't have broken it.
> 
> It should have. `compute_relation()` function was moved from `nfa.hh` to `nfa-algorithms.hh`. Therefore, we try to bind to undefined method `compute_relation()` in `nfa.hh` in Python binding.

You misunderstood the `shouldn't have` :-). I meant PR that focuses on improving cmake, should not in ideal world include changes to API. But, Martin explained why it was necessary.

**Comment by tfiedor** [comment]
@martinhruska fixed

**Comment by martinhruska** [comment]
Great, thank you, I'm gonna merge it.

**Pull request merged**